Sj/ga migrate linting

SJaffa opened the pull request:
Check the code complies with C++ standards and makes changes where appropriate.

[This similar example](https://github.com/marketplace/actions/clang-format-lint) suggests adding a step to commit the suggested changes. Do we want to do that or just prompt the user to commit?

SJaffa committed:
Linting GHA on latest Ubuntu

SJaffa committed:
Changes name of jobs and stages to be more sensible

SJaffa committed:
Added warning to commit lint changes if problems are detected

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+          sudo apt-get update -y -qq 
+          sudo apt-get install clang-format
+          
+      - name: Configure
+        run: | 
+          mkdir build
+          cd build
+          cmake .. \
+          -Ddompi=OFF \
+          -Dopenmp=OFF \
+          -Ddocs=OFF 
```
Do we need this step? It seems like the steps below can run fine without this step and it is tested separately in the `cmake.yml` workflow. If we don't need this, I think we can also remove the `Install dependencies` step.

SJaffa replied:
```suggestion

```

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+          curl https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+          echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal main" | sudo tee -a /etc/apt/sources.list
+          sudo apt-get update -y -qq 
+          sudo apt-get install clang-format
```
```suggestion
          sudo apt update -y -qq 
          sudo apt install clang-format
```

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+      - name: Checkout repo
+        uses: actions/checkout@v2
+        
+      - run: sudo apt-get update
```
It's not a big deal, but for example [this answer](https://askubuntu.com/questions/445384/what-is-the-difference-between-apt-and-apt-get) suggests using `apt` rather than `apt-get`
```suggestion
      - run: sudo apt update
```

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+  
+jobs:
+  linting:
+    runs-on: ubuntu-latest
```
I think it's better to fix the ubuntu version. 22.04 is not that far away 😬 
```suggestion
    runs-on: ubuntu-20.04
```

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+
+on:
+  push:
+    branches: [ sj/ga-migrate-linting ]
```
Updated the branch name. Didn't we decide to run this on pull request rather than push?
```suggestion
on:
  push:
    branches: [ development ]
```

tkoskela reviewed (changes requested):
Thanks Sarah, this is looking pretty good! My main question is why do we need to install the dependencies and run the cmake configure if all we are doing is inspecting the source code? Did I misunderstand what the linter is doing?

SJaffa committed:
Update .github/workflows/linting.yml

Co-authored-by: Tuomas Koskela <t.koskela@ucl.ac.uk>

Apply suggestions from code review

Small improvements

Co-authored-by: Tuomas Koskela <t.koskela@ucl.ac.uk>

Update .github/workflows/linting.yml

Run on push and pull

Apply suggestions from code review

Remove unnecessary steps

Co-authored-by: Tuomas Koskela <t.koskela@ucl.ac.uk>

Apply suggestions from code review

Add notes to clarify for future users

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+    
+      - name: Extra dependencies
+        run: |
+          curl https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+          echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal main" | sudo tee -a /etc/apt/sources.list
+          sudo apt update -y -qq 
```
We can install `clang-format` from the default apt repo, this seems unnecessary
```suggestion

```

SJaffa commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,43 @@
+
+on:
+  push:
+    branches: [ development ]
```
```suggestion
    branches: [ development ]
  pull-request:
    branches: [development]    
```

SJaffa commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,45 @@
+      - name: Checkout repo
+        uses: actions/checkout@v2
+        
+      - run: sudo apt update
+    
+      - name: Install dependencies
+        run: sudo apt install libfftw3-dev libtiff5-dev libboost-all-dev libeigen3-dev libyaml-cpp-dev libcfitsio-dev casacore-dev libllvm7 
```
```suggestion

```

SJaffa commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,30 @@
+
+      - name: Run linting
+        run: | 
+          find ../ -regex '.*\.\(cc\|h\)' -not -iname '*.in.h' | xargs -I{} -P 10 clang-format -i -style=file {}; git diff
```
```suggestion
          find ../ -regex '.*\.\(cc\|h\)' -not -iname '*.in.h' | xargs -I{} -P 10 clang-format -i -style=Google {}; git diff
```

tkoskela replied:
@dpshelio : Keep using `file`, we have some custom changes to the style and we want to keep control of the version to ourselves.

The file is in the root of the repo https://github.com/astro-informatics/purify/blob/development/.clang-format

SJaffa commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,30 @@
+        run: |
+          sudo apt install clang-format
+          
+
```
```suggestion
# Check code meets Google C++ style guide https://google.github.io/styleguide/cppguide.html
```

SJaffa commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,30 @@
+        uses: actions/checkout@v2
+        
+    
+      - name: Extra dependencies
```
```suggestion
      - name: Install dependencies
```
